### PR TITLE
Setting up build type and debug compiler flags

### PR DESCRIPTION
**Pull request by sergeyklay**
Hello,

### Proposal

This PR adds/changes the following things:

1. Setting build type to 'Debug' when no build type was specified by user (see `Re2cBootstrapBuildType.cmake`), because the default build type is an _empty string_
1. Verify build type (see `Re2cBootstrapBuildType.cmake`) to use only known ad supported types
1. Move flags related stuff to common place (see `Re2cBootstrapCompilerFlags.cmake`) to keep CMakeLists.txt clean
1. Output used compiler flags (see `Re2cBootstrapCompilerFlags.cmake`) to verify default ones

The four build types that CMake includes out of the box are `Release`, `Debug`, `RelWithDebInfo`, and `MinSizeRel`. Correspondingly, CMake ships with default values for each defined build type in `CMAKE_CXX_FLAGS_<buildType>`. Thus, right now we have the following _default_ flags:

|                           | Debug                                        | Release                                  |
| :---                     |     ---:                                          |          ---:                                |
| **GCC/Clang**  | `-g`                                              | `-O3 -DNDEBUG`                   |
| **MSVC**          |  `/MDd /Zi /Ob0 /Od /RTC1`       | `/MD /O2 /Ob2 /DNDEBUG` |

Also I removed the following hardcoded flag:
  ```cmake
  if(NOT CMAKE_BUILD_TYPE)
       try_cxxflag("-O2")
  endif()
  ```
because this one should be set by build type.

### Questions

1. Should we use `Release` build type by default?
2. Do we need append to `CMAKE_CXX_FLAGS_<buildType>` some custom flags?
3. Should I remove redundant check `if(CMAKE_BUILD_TYPE IN_LIST ALLOWABLE_BUILD_TYPES)` (see `Re2cBootstrapCompilerFlags.cmake`) due to the fact that now the configuration will always be set?

### Notes

This PR reveals some "meaninglessness" of the Travis' matrix configuration (see COMMON_FLAGS):
https://github.com/skvadrik/re2c/blob/c23ed638c933219060c52bb0f0733930b73050aa/.travis.yml#L21-L25

I also plan to modify how the `try_cxxflag` function works to get rid of bad practices, however I'd prefer to do this a bit later.

### CODE REVIEW

**Comment by skvadrik** [comment]
Thanks! I have a few suggestions:

- Let's rename cmake/Re2cBootstrapCompilerFlags.cmake -> cmake/Re2cCompilerFlags.cmake and cmake/Re2cBootstrapBuildType.cmake -> cmake/Re2cBuildType.cmake.
- Maybe merge cmake/Re2cTryCXXFlag.cmake contents into cmake/Re2cCompilerFlags.cmake?

> Should we use Release build type by default?

Yes.

> Do we need append to CMAKE_CXX_FLAGS_<buildType> some custom flags?

`-DRE2C_DEBUG` for debug builds and a few other defines, currently added here: https://github.com/skvadrik/re2c/blob/c23ed638c933219060c52bb0f0733930b73050aa/CMakeLists.txt#L116-L121 It is not necessary to move them, but feel free if you think it's cleaner.

> Should I remove redundant check if(CMAKE_BUILD_TYPE IN_LIST ALLOWABLE_BUILD_TYPES) (see Re2cBootstrapCompilerFlags.cmake) due to the fact that now the configuration will always be set?

Yes, I'd remove it. Maybe `ALLOWABLE_BUILD_TYPES` is not needed, if the build type is always set and cmake outputs a reasonable error message for inexistent build types.

> This PR reveals some "meaninglessness" of the Travis' matrix configuration (see COMMON_FLAGS)

Why is it meaningless? `COMMON_FLAGS` always set `-O2 -g` , except for release build. Debug build with `-O2` is the type of build used for development: it has debug symbols and debug checks/asserts, but it isoptimized so that the tests run fast (cmake has `RelWithDebInfo`, but there was something that prevented me from using it -- I forgot what). I don't see much sense in testing purely debug build (without `-O0`), because it's just slower than debug build with `-O2` but has all the same debug checks.

**Comment by sergeyklay** [comment]
Done :) Thank you for the feedback!

---

> Yes, I'd remove it. Maybe `ALLOWABLE_BUILD_TYPES` is not needed, if the build type is always set and cmake outputs a reasonable error message for inexistent build types.

 `ALLOWABLE_BUILD_TYPES` is needed because everyone can run cmake as follows:
```bash
$ cmake -S . \
    -B cmake-build \
    -DCMAKE_BUILD_TYPE=FooBarBaz
```

What is FooBarBaz? What flags should we use?


There are strategies when, depending on configuration you may want to set specific flag inside CMake lists:

```cmake
target_compile_options(
  some-target-name
  INTERFACE
      # Baseline reasonable warnings for MSVC
      $<$<CXX_COMPILER_ID:MSVC>:$<$<CONFIG:Debug>:/W4>>

      # Baseline reasonable warnings for CLang/GCC
      $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:
      $<$<BOOL:${WARNINGS_AS_ERRORS}>:-Wall>>
)
```

But now, with FooBarBaz, you know nothing.




> `COMMON_FLAGS` always set `-O2 -g` , except for release build.


I meant optimization (`-O`) is set internally by CMake depends on build type and no need to set it by hand, except when the default level of `-O` is not enough for you. I'm not sure if this significantly speeds up the build, but of course we can leave that as it is. I'm OK with this. Nevertheless `-g` is not needed in `COMMON_FLAGS`, because CMake always sets `-g` internally for Debug configuration.

**Comment by sergeyklay** [comment]
I also added verbose output about the flags used. This may help in future to debug used flags:

```text
-- Common compiler flags for C: 
-- Debug compiler flags for C: -g
-- Common compiler flags for CXX: -fdiagnostics-color=always -Werror=return-type -Wsign-conversion -Wconversion -Wsuggest-attribute=format -Wredundant-decls -Wformat=2 -pedantic -Weffc++ -Wextra -Wall -W 
-- Debug compiler flags for CXX: -g
```

It is interesting that for Windows we have three languages:

```text
-- Common compiler flags for C: /DWIN32 /D_WINDOWS /W3
-- Debug compiler flags for C: /MDd /Zi /Ob0 /Od /RTC1
-- Common compiler flags for CXX: -Wall /DWIN32 /D_WINDOWS /W3 /GR /EHsc
-- Debug compiler flags for CXX: /MDd /Zi /Ob0 /Od /RTC1
-- Common compiler flags for RC: -DWIN32
-- Debug compiler flags for RC: -D_DEBUG
```

And as we can see only `-Wall` is supported by Visual C++. However there are native `/Wall` flag. The Visual C++ `/Wall` enables all of the warnings that are disabled by default at `/W4`. But that's another story and I'll deal with it later :)

**Review by skvadrik — COMMENTED** [review]
Thinking more about default build type, I actually recall one reason why this might be problematic (sorry I couldn't remember it earlier). 

As I understand, setting default build type means that CMake will always append some compiler flags that are impossible to override. For example, it is impossible to build with `-O2` but without `-g`, because `Debug` will always append `-g` and `Release` will always append `-O3`. This is just one example, but the general idea is that choosing one of the preset build types forces some flags without a way to override them.

That's why it might be a better idea to have empty build type by default, but have some sane defaults set in CMakeLists.txt (`-O2 -g`) that can be overridden by the user (currently they can because `try_cxxflag` *pre*pends flags, thus any user-defined flags passed in `-DCMAKE_CXX_FLAGS` will go after default flags and will take precedence).

**Comment by sergeyklay** [comment]
Well, I think the general idea is not to force user to play with flags and dealing which one is better to compiler program (`-O2`, `-O3`, `-Os`, etc). This is why there are build types (e.g. configuration names). All the user needs to know is the use of a named configuration. However, even this is not necessary if the user just wants to compile the program.

As for the impossibility of redefinition - I strongly doubt it. At least I have never heard of such behavior. AFAIK, if we have the following flags together `<some other flags> ...  -O1 -O2 -O3 <some other flags> ...` , then the last one has the highest priority. Can you share a reference that describes impossibility of redefinition in case of using CMake build types?

File: cmake/Re2cBuildType.cmake [hunk 1]
```diff
@@ -0,0 +1,11 @@
+set(_allowable_build_types Debug Release RelWithDebInfo MinSizeRel)
+
+# Specify the build type
+if(NOT CMAKE_BUILD_TYPE)
+    message(STATUS "Setting build type to 'Release' as none was specified.")
+    set(CMAKE_BUILD_TYPE Debug CACHE STRING "Choose the type of build." FORCE)
```
[the remark this answers is not in the document]
Comment (sergeyklay) [reply cmake/Re2cBuildType.cmake:6]: Good catch. I'll revisit this a bit later. Thank you!

**Comment by skvadrik** [comment]
> Well, I think the general idea is not to force user to play with flags

For the most users CMake defaults should be ok. But I have in mind someone like a distro maintainer trying to rebuild re2c after a compiler upgrade, hitting a bug that reproduces with some weird flag combination. In such cases it is useful to be able to pass user-defined `CXXFLAGS` in some standard way, without patching the build files.

> Can you share a reference that describes impossibility of redefinition in case of using CMake build types?

I don't really have an official doc ([here](https://github.com/spack/spack/issues/11223) is some related discussion). My understanding is that the standard way to pass user-defined `CXXFLAGS` in CMake is to define `CMAKE_CXX_FLAGS` variable. The problem is, in the final flag sequence `CMAKE_CXX_FLAGS` comes before build flavour specific variables `CMAKE_CXX_FLAGS_RELEASE`, `CMAKE_CXX_FLAGS_DEBUG` and so on. Therefore if a non-empty build type is used, CMake will append some flags after the user-defined ones and override them.

You can see this if you try to build re2c with `-O2`, but without `-g`. With your PR it is hard to achieve, because CMake either appends `-O3` in release builds, or `-g` in debug builds. It is possible to override with `-DCMAKE_CXX_FLAGS_RELEASE="-O2"`, but this has the drawback that it resets all CMake-defined release flags, and requires special CMake knowledge.

I understand that it is perhaps the "CMake way" to keep to some predefined configurations that are known to work and well-tested on all platforms, and I have seen an [email thread](https://cmake.org/pipermail/cmake/2006-February/008371.html) suggesting that empty build type is a bad idea. Therefore I'm leaning towards setting `-DCMAKE_BUILD_TYPE=Release` by default. But I still think this is a bad design decision in CMake, and user-defined flags should always take priority.

**Comment by skvadrik** [comment]
> I also added verbose output about the flags used.

This is very useful!

**Comment by sergeyklay** [comment]
Done. Now there are only few changes:
- Verify build type (if provided)
- Verbose output about the flags used
- Moved flags-related stuff to the common place

**Comment by skvadrik** [comment]
Merged, thank you!

**Comment by sergeyklay** [comment]
You're welcome :)

**Pull request merged**